File: trunk/+tom/DynamicModelDefault.m
classdef DynamicModelDefault < tom.DynamicModel
  
  properties (GetAccess = protected, SetAccess = protected)
    interval
  end
  
  methods (Static = true, Access = public)
    function initialize(name)
      function text = componentDescription
        text = 'This default dynamic model represents a stationary body at the world origin with no input parameters.';
      end
      tom.DynamicModel.connect(name, @componentDescription, @tom.DynamicModelDefault);
    end
  end
  
  methods (Access = public, Static = true)
    function this = DynamicModelDefault(initialTime, uri)
      this = this@tom.DynamicModel(initialTime, uri);
      this.interval = tom.TimeInterval(initialTime, tom.WorldTime(Inf));
    end
  end
  
  methods (Access = public, Static = false)   
    function num = numInitialLogical(this)
      assert(isa(this, 'tom.DynamicModel'));
      num = uint32(0);
    end
    
    function num = numInitialUint32(this)
      assert(isa(this, 'tom.DynamicModel'));
      num = uint32(0);
    end
  
    function num = numExtensionLogical(this)
      assert(isa(this, 'tom.DynamicModel'));
      num = uint32(0);
    end
    
    function num = numExtensionUint32(this)
      assert(isa(this, 'tom.DynamicModel'));
      num = uint32(0);
    end

    function num = numExtensionBlocks(this)
      assert(isa(this, 'tom.DynamicModel'));
      num = uint32(0);
    end
    
    function v = getInitialLogical(this, p)
      assert(isa(this, 'tom.DynamicModel'));
      assert(isa(p, 'uint32'));
      v = false;
      assert(isa(v, 'logical'));
      error('The default dynamic model has no input parameters.');
    end

    function v = getInitialUint32(this, p)
      assert(isa(this, 'tom.DynamicModel'));
      assert(isa(p, 'uint32'));
      v = uint32(0);
      assert(isa(v, 'uint32'));
      error('The default dynamic model has no input parameters.');
    end

    function v = getExtensionLogical(this, b, p)
      assert(isa(this, 'tom.DynamicModel'));
      assert(isa(b, 'uint32'));
      assert(isa(p, 'uint32'));
      v = false;
      assert(isa(v, 'logical'));
      error('The default dynamic model has no input parameters.');
    end

    function v = getExtensionUint32(this, b, p)
      assert(isa(this, 'tom.DynamicModel'));
      assert(isa(b, 'uint32'));
      assert(isa(p, 'uint32'));
      v = uint32(0);
      assert(isa(v, 'uint32'));
      error('The default dynamic model has no input parameters.');
    end

    function setInitialLogical(this, p, v)
      assert(isa(this, 'tom.DynamicModel'));
      assert(isa(p, 'uint32'));
      assert(isa(v, 'logical'));
      error('The default dynamic model has no input parameters.');
    end

    function setInitialUint32(this, p, v)
      assert(isa(this, 'tom.DynamicModel'));
      assert(isa(p, 'uint32'));
      assert(isa(v, 'uint32'));
      error('The default dynamic model has no input parameters.');
    end
    
    function setExtensionLogical(this, b, p, v)
      assert(isa(this, 'tom.DynamicModel'));
      assert(isa(b, 'uint32'));
      assert(isa(p, 'uint32'));
      assert(isa(v, 'logical'));
      error('The default dynamic model has no input parameters.');
    end
    
   function setExtensionUint32(this, b, p, v)
      assert(isa(this, 'tom.DynamicModel'));
      assert(isa(b, 'uint32'));
      assert(isa(p, 'uint32'));
      assert(isa(v, 'uint32'));
      error('The default dynamic model has no input parameters.');
    end
    
    function cost = computeInitialBlockCost(this)
      assert(isa(this, 'tom.DynamicModel'));
      cost = 0;
      assert(isa(cost, 'double'));
    end

    function cost = computeExtensionBlockCost(this, b)
      assert(isa(this, 'tom.DynamicModel'));
      assert(isa(b, 'uint32'));
      cost = 0;
      assert(isa(cost, 'double'));
      error('The default dynamic model has no extension blocks.');
    end
    
    function interval = domain(this)
      interval = this.interval;
    end
  
    function pose = evaluate(this, t)
      pose.p = [6378137.0; 0; 0];
      pose.q = [1; 0; 0; 0];
      pose = tom.Pose(pose);
      pose = repmat(pose, [1, numel(t)]);
      for k = find(t<this.interval.first)
        pose(k) = tom.Pose;
      end
    end

    function tangentPose = tangent(this, t)
      tangentPose.p = [6378137.0; 0; 0];
      tangentPose.q = [1; 0; 0; 0];
      tangentPose.r = [0; 0; 0];
      tangentPose.s = [0; 0; 0];
      tangentPose = tom.TangentPose(tangentPose);
      tangentPose = repmat(tangentPose, [1, numel(t)]);
      for k = find(t<this.interval.first)
        tangentPose(k) = tom.TangentPose;
      end
    end
    
    function extend(this)
      assert(isa(this, 'tom.DynamicModel'));
    end
    
    function obj = copy(this)
      obj = tom.DynamicModelDefault(tom.WorldTime(0),'');
      obj.interval = this.interval;
    end
  end
  
end

File: trunk/+tom/DynamicModelDefault.cpp
#include "DynamicModel.h"
#include "WGS84.h"

namespace tom
{
  /** This default dynamic model represents a stationary body at the world origin with no input parameters. */
  class DynamicModelDefault : public virtual DynamicModel
  {
  protected:
    std::pair<double, double> interval; /**< stores the Trajectory domain */

  public:
    /** Constructor and parent class initializer of the same form. */
    DynamicModelDefault(const double initialTime, const std::string uri) :
      DynamicModel(initialTime, uri)
    {
      interval.first = initialTime;
      interval.second = INF;
      return;
    }

    uint32_t numInitial(void)
    {
      return (0);
    }

    uint32_t numExtension(void)
    {
      return (0);
    }

    uint32_t numBlocks(void)
    {
      return (0);
    }

    uint32_t getInitial(uint32_t parameterIndex)
    {
      throw("The default dynamic model has no input parameters");
      return (0);
    }

    uint32_t getExtension(uint32_t blockIndex, uint32_t parameterIndex)
    {
      throw("The default dynamic model has no input parameters");
      return (0);
    }

    void setInitial(uint32_t parameterIndex, uint32_t value)
    {
      throw("The default dynamic model has no input parameters");
      return;
    }

    void setExtension(uint32_t blockIndex, uint32_t parameterIndex, uint32_t value)
    {
      throw("The default dynamic model has no input parameters");
      return;
    }

    double computeInitialCost(void)
    {
      return (0.0);
    }

    double computeExtensionCost(uint32_t blockIndex)
    {
      throw("The default dynamic model has no extension blocks.");
      return (0.0);
    }

    std::pair<double, double> domain(void)
    {
      return (interval);
    }

    void evaluate(const double& time, Pose& pose)
    {
      static const Pose nullPose;
      if(time<interval.first)
      {
        pose = nullPose;
      }
      else
      {
        pose.p[0] = WGS84::majorRadius;
        pose.p[1] = 0.0;
        pose.p[2] = 0.0;
        pose.q[0] = 1.0;
        pose.q[1] = 0.0;
        pose.q[2] = 0.0;
        pose.q[3] = 0.0;
      }
      return;
    }

    void tangent(const double& time, TangentPose& tangentPose)
    {
      static const TangentPose nullTangentPose;
      if(time<interval.first)
      {
        tangentPose = nullTangentPose;
      }
      else
      {
        tangentPose.p[0] = WGS84::majorRadius;
        tangentPose.p[1] = 0.0;
        tangentPose.p[2] = 0.0;
        tangentPose.q[0] = 1.0;
        tangentPose.q[1] = 0.0;
        tangentPose.q[2] = 0.0;
        tangentPose.q[3] = 0.0;
        tangentPose.r[0] = 0.0;
        tangentPose.r[1] = 0.0;
        tangentPose.r[2] = 0.0;
        tangentPose.s[0] = 0.0;
        tangentPose.s[1] = 0.0;
        tangentPose.s[2] = 0.0;
      }
      return;
    }

    void extend(void)
    {
      return;
    }

    tom::DynamicModel* copy(void)
    {
      double initialTime = this->interval.first;
      std::string uri = "";
      BrownianPlanar* obj = new BrownianPlanar(initialTime, uri);
      obj->interval = this->interval;
      return (obj);
    }

  private:
    static std::string componentDescription(void)
    {
      return ("This default dynamic model represents a stationary body at the world origin with no input parameters.");
    }

    static DynamicModel* componentFactory(const double initialTime, const std::string uri)
    {
      return (new DynamicModelDefault(initialTime, uri));
    }

  protected:
    static void initialize(std::string name)
    {
      connect(name, componentDescription, componentFactory);
    }
    friend class DynamicModelDefaultInitializer;
  };

  /** This class initializes DynamicModelDefault before the execution of main(). */
  class DynamicModelDefaultInitializer
  {
  public:
    DynamicModelDefaultInitializer(void)
    {
      DynamicModelDefault::initialize("tom");
    }
  } _DynamicModelDefaultInitializer;
}
